src/java/org/apache/cassandra/index/sai/plan/Expression.java

```diff
@@ -333,22 +335,70 @@ public boolean isSatisfiedBy(ByteBuffer columnValue)
         return true;
     }
 
-    public ByteBuffer getLowerBound()
+    /**
+     * Returns the lower bound of the expression as a ByteComparable with an encoding based on the version, the
+     * validator, and whether the expression is in memory or on disk.
+     * @param version the version of the index
+     * @param inMemory whether the expression is in memory or on disk
+     * @return
+     */
+    public ByteComparable getEncodedLowerBoundByteComparable(Version version, boolean inMemory)
+    {
+        var unencodedLowerBound = getLowerBoundByteBuffer(version);
+        if (unencodedLowerBound == null)
+            return null;
+        var terminator = lower.inclusive ? ByteSource.TERMINATOR : ByteSource.GT_NEXT_COMPONENT;
+        return getBoundByteComparable(unencodedLowerBound, version, inMemory, terminator);
+    }
+
+    /**
+     * Returns the upper bound of the expression as a ByteComparable with an encoding based on the version, the
+     * validator, and whether the expression is in memory or on disk.
+     * @param version the version of the index
+     * @param inMemory whether the expression is in memory or on disk
+     * @return
+     */
+    public ByteComparable getEncodedUpperBoundByteComparable(Version version, boolean inMemory)
+    {
+        var unencodedUpperBound = getUpperBoundByteBuffer(version);
+        if (unencodedUpperBound == null)
+            return null;
+        var terminator = upper.inclusive ? ByteSource.TERMINATOR : ByteSource.LT_NEXT_COMPONENT;
+        return getBoundByteComparable(unencodedUpperBound, version, inMemory, terminator);
+    }
+
+    private ByteComparable getBoundByteComparable(ByteBuffer unencodedBound, Version version, boolean inMemory, int terminator)
+    {
+        if (TypeUtil.isComposite(validator) && version.onOrAfter(Version.DB))
+            // Note that for ranges that have one unrestricted bound, we technically do not need the terminator
+            // because we use the 0 or the 1 at the end of the first component as the bound. However, it works
+            // with the terminator, so we use it for simplicity.
+            return TypeUtil.asComparableBytes(unencodedBound, terminator, (CompositeType) validator);
+        else if (inMemory)
+            return version.onDiskFormat().encodeForInMemoryTrie(unencodedBound, validator);
+        else
+            return version.onDiskFormat().encodeForOnDiskTrie(unencodedBound, validator);
+    }
+
+    public ByteBuffer getLowerBoundByteBuffer(Version version)
```

Comment:
nit: ByteBuffer in the name of the method is redundant. `getLowerBound` would be enough; the type signature already says it returns a `ByteBuffer`

Reply:
Good point. I changed the name because I found myself confused about what I had when. I went with `getPartiallyEncodedLowerBound` and added comments to indicate why.